genesis/engine/entities/rigid_entity/rigid_entity.py

```diff
@@ -1970,32 +1975,38 @@ def set_quat(self, quat, envs_idx=None, *, relative=False, zero_velocity=True, u
     @gs.assert_built
     def get_verts(self):
         """
-        Get the all vertices of the entity (using collision geoms).
+        Get the all vertices of the entity based on collision geometries.
 
         Returns
         -------
-        verts : torch.Tensor, shape (n_verts, 3) or (n_envs, n_verts, 3)
-            The vertices of the entity (using collision geoms).
+        verts : torch.Tensor, shape (n_envs, n_verts, 3)
+            The vertices of the entity.
         """
-        self._update_verts_for_geom()
-        if self.is_free:
-            tensor = torch.empty(
-                self._solver._batch_shape((self.n_verts, 3), True), dtype=gs.tc_float, device=gs.device
-            )
-            _kernel_get_free_verts(tensor, self._verts_state_start, self.n_verts, self._solver.free_verts_state)
-            if self._solver.n_envs == 0:
-                tensor = tensor.squeeze(0)
-        else:
-            tensor = torch.empty((self.n_verts, 3), dtype=gs.tc_float, device=gs.device)
-            _kernel_get_fixed_verts(tensor, self._verts_state_start, self.n_verts, self._solver.fixed_verts_state)
-        return tensor
-
-    @gs.assert_built
-    def _update_verts_for_geom(self):
         for i_g_ in range(self.n_geoms):
             i_g = i_g_ + self._geom_start
             self._solver.update_verts_for_geom(i_g)
 
+        n_free_verts, n_fixed_verts = 0, 0
+        for link in self.links:
+            if link.is_fixed:
+                n_fixed_verts += link.n_verts
+            else:
+                n_free_verts += link.n_verts
+        n_verts = n_free_verts + n_fixed_verts
+
+        tensor = torch.empty(self._solver._batch_shape((n_verts, 3), True), dtype=gs.tc_float, device=gs.device)
+        _kernel_get_fixed_verts(
+            tensor[0, :n_fixed_verts], self._fixed_verts_state_start, n_fixed_verts, self._solver.fixed_verts_state
+        )
+        tensor[1:, :n_fixed_verts] = tensor[0, :n_fixed_verts]
+        _kernel_get_free_verts(
+            tensor[:, n_fixed_verts:], self._free_verts_state_start, n_free_verts, self._solver.free_verts_state
```

Comment:
why is fixed verts always before free verts? Is this consistent with face index?

Reply:
Probably not. I will update the PR to preserve geometry ordering (and therefore face index).